homeassistant/components/tplink/sensor.py

```diff
@@ -144,38 +79,35 @@ async def async_setup_entry(
     entities: list[CoordinatedTPLinkEntity] = []
     device = parent_coordinator.device
 
-    for idx, child in enumerate(device.children):
-        # TODO: nicer way for multi-coordinator updates.
-        if isinstance(device, SmartStrip):
-            entities.extend(
-                _async_sensors_for_device(
-                    child, children_coordinators[idx], parent=device
-                )
-            )
-        else:
-            entities.extend(
-                _entities_for_device(
-                    child,
-                    feature_type=Feature.Sensor,
-                    entity_class=Sensor,
-                    coordinator=parent_coordinator,
-                    parent=device,
-                )
-            )
-
     entities.extend(
-        _entities_for_device(
+        _entities_for_device_and_its_children(
             device,
             feature_type=Feature.Sensor,
-            entity_class=Sensor,
+            entity_class=TPLinkSensor,
             coordinator=parent_coordinator,
+            extra_filter=_sensor_filter,
+            children_coordinators=children_coordinators,
         )
     )
 
     async_add_entities(entities)
 
 
-class Sensor(CoordinatedTPLinkEntity, SensorEntity):
+def _sensor_filter(device: Device, feature: Feature) -> bool:
+    "Filter sensors."
+    # today's consumption not available, when device was off all the day
+    if feature.id == ATTR_TODAY_ENERGY_KWH:
+        # iot bulb's do not report this information, so filter it out
+        if isinstance(device, IotBulb):
+            return False
+        return True
+    # By default do not show emeter features with None values
+    if feature.id in EMETER_FEATURE_IDS:
+        return feature.value is not None
+    return True
```

Comment:
We should move this sort of handling out from the integration. IotBulb should not create a feature for data we know won't exist, and we probably should not create features for non-existing data (except in specific cases?).

Reply:
This is now addressed in the library in https://github.com/python-kasa/python-kasa/pull/976